src/coreclr/src/gc/gc.cpp

```diff
@@ -35103,6 +35134,64 @@ HRESULT GCHeap::Initialize()
 
 #ifdef HOST_64BIT
     gc_heap::heap_hard_limit = (size_t)GCConfig::GetGCHeapHardLimit();
+    gc_heap::heap_hard_limit_soh = (size_t)GCConfig::GetGCHeapHardLimitSOH();
+    gc_heap::heap_hard_limit_loh = (size_t)GCConfig::GetGCHeapHardLimitLOH();
+    gc_heap::heap_hard_limit_poh = (size_t)GCConfig::GetGCHeapHardLimitPOH();
+
+    if (gc_heap::heap_hard_limit_soh || gc_heap::heap_hard_limit_loh || gc_heap::heap_hard_limit_poh)
+    {
+        if (!gc_heap::heap_hard_limit_soh)
+        {
+            return E_INVALIDARG;
+        }
+        if (!gc_heap::heap_hard_limit_loh)
+        {
+            return E_INVALIDARG;
+        }
+        if (gc_heap::heap_hard_limit_poh < min_segment_size_hard_limit)
+        {
+            gc_heap::heap_hard_limit_poh = min_segment_size_hard_limit;
+        }
+        // This tells the system there is a hard limit, but otherwise we will not compare against this value.
+        gc_heap::heap_hard_limit = 1;
+    }
+    else
+    {
+        uint32_t percent_of_mem_soh = (uint32_t)GCConfig::GetGCHeapHardLimitSOHPercent();
+        uint32_t percent_of_mem_loh = (uint32_t)GCConfig::GetGCHeapHardLimitLOHPercent();
+        uint32_t percent_of_mem_poh = (uint32_t)GCConfig::GetGCHeapHardLimitPOHPercent();
+        if (percent_of_mem_soh || percent_of_mem_loh || percent_of_mem_poh)
+        {
+            if ((percent_of_mem_soh <= 0) || (percent_of_mem_soh >= 100))
+            {
+                return E_INVALIDARG;
+            }
+            if ((percent_of_mem_loh <= 0) || (percent_of_mem_loh >= 100))
+            {
+                return E_INVALIDARG;
+            }
+            else if ((percent_of_mem_poh < 0) || (percent_of_mem_poh >= 100))
+            {
+                return E_INVALIDARG;
+            }
+            if (percent_of_mem_soh + percent_of_mem_loh + percent_of_mem_poh >= 100)
+            {
```

Comment:
nit - GC coding style calls for () here -
`if ((percent_of_mem_soh + percent_of_mem_loh + percent_of_mem_poh) >= 100)`